File: code/backend/shared/src/main/java/com/nxp/iemdm/shared/intf/operational/landingai/ConfusionMatrixService.java
package com.nxp.iemdm.shared.intf.operational.landingai;

import com.nxp.iemdm.shared.dto.landingai.CellDetailResponse;
import com.nxp.iemdm.shared.dto.landingai.ConfusionMatrixResponse;
import com.nxp.iemdm.shared.dto.landingai.GroundTruthLabelDTO;
import com.nxp.iemdm.shared.dto.landingai.ImageWithLabelsDTO;
import com.nxp.iemdm.shared.dto.landingai.PredictionLabelDTO;
import java.util.List;

/** Service interface for confusion matrix operations. */
public interface ConfusionMatrixService {

  /**
   * Calculate confusion matrix for a model and evaluation set.
   *
   * @param modelId Model ID
   * @param evaluationSet Evaluation set (TRAIN, DEV, TEST)
   * @return Confusion matrix response with grid, classes, and metrics
   */
  ConfusionMatrixResponse calculateConfusionMatrix(Long modelId, String evaluationSet);

  /**
   * Get detail for a specific cell (GT×Pred combination).
   *
   * @param modelId Model ID
   * @param evaluationSet Evaluation set (TRAIN, DEV, TEST)
   * @param gtClassId Ground truth class ID
   * @param predClassId Prediction class ID
   * @return Cell detail response with images
   */
  CellDetailResponse getCellDetail(
      Long modelId, String evaluationSet, Long gtClassId, Long predClassId);

  /**
   * Get all images in evaluation set with correctness indicators.
   *
   * @param modelId Model ID
   * @param evaluationSet Evaluation set (TRAIN, DEV, TEST)
   * @return List of images with ground truth and prediction labels
   */
  List<ImageWithLabelsDTO> getAllImages(Long modelId, String evaluationSet);

  /**
   * Get prediction labels for a model filtered by evaluation set. Backend 根據 evaluationSet 參數過濾資料，
   * 只回傳指定 evaluation set 的資料。 Requirements: 26.5, 31.1, 35.3
   *
   * @param modelId Model ID
   * @param evaluationSet Evaluation set (TRAIN, DEV, TEST)
   * @return List of prediction labels for the specified evaluation set
   */
  List<PredictionLabelDTO> getPredictionLabels(Long modelId, String evaluationSet);

  /**
   * Get ground truth labels for a model's project filtered by evaluation set. Backend 根據
   * evaluationSet 參數過濾資料，只回傳指定 evaluation set 的資料。 Requirements: 26.5, 31.1, 35.3
   *
   * @param modelId Model ID (用來找到對應的 snapshot)
   * @param evaluationSet Evaluation set (TRAIN, DEV, TEST)
   * @return List of ground truth labels for the specified evaluation set
   */
  List<GroundTruthLabelDTO> getGroundTruthLabels(Long modelId, String evaluationSet);
}

File: code/frontend/src/main/frontend/src/app/services/landingai/confusion-matrix.service.ts
import { HttpClient, HttpErrorResponse } from "@angular/common/http";
import { Injectable } from "@angular/core";
import { Observable, throwError } from "rxjs";
import { catchError } from "rxjs/operators";
import { Configuration } from "app/utils/configuration";
import {
  ConfusionMatrixData,
  CellDetailResponse,
  ImageWithLabels,
} from "app/models/landingai/confusion-matrix.model";

/**
 * Service for confusion matrix operations
 * Calls backend API endpoints for confusion matrix data
 */
@Injectable({
  providedIn: "root",
})
export class ConfusionMatrixService {
  private readonly actionUrl: string;

  constructor(
    private http: HttpClient,
    private configuration: Configuration
  ) {
    this.actionUrl =
      this.configuration.ServerWithApiUrl + "landingai/confusion-matrix/";
  }

  /**
   * Get confusion matrix data for a model and evaluation set
   * @param modelId The model ID
   * @param evaluationSet The evaluation set (train, dev, test)
   * @returns Observable of ConfusionMatrixData
   */
  public getConfusionMatrix(
    modelId: number,
    evaluationSet: "train" | "dev" | "test"
  ): Observable<ConfusionMatrixData> {
    return this.http
      .get<ConfusionMatrixData>(`${this.actionUrl}${modelId}/${evaluationSet}`)
      .pipe(catchError(this.handleError));
  }

  /**
   * Get cell detail (images for specific GT×Pred combination)
   * @param modelId The model ID
   * @param evaluationSet The evaluation set (train, dev, test)
   * @param gtClassId Ground truth class ID
   * @param predClassId Prediction class ID
   * @returns Observable of CellDetailResponse
   */
  public getCellDetail(
    modelId: number,
    evaluationSet: "train" | "dev" | "test",
    gtClassId: number,
    predClassId: number
  ): Observable<CellDetailResponse> {
    const params = {
      gtClassId: gtClassId.toString(),
      predClassId: predClassId.toString(),
    };

    return this.http
      .get<CellDetailResponse>(
        `${this.actionUrl}${modelId}/${evaluationSet}/cell`,
        { params }
      )
      .pipe(catchError(this.handleError));
  }

  /**
   * Get all images in evaluation set with labels and predictions
   * @param modelId The model ID
   * @param evaluationSet The evaluation set (train, dev, test)
   * @returns Observable of ImageWithLabels array
   */
  public getAllImages(
    modelId: number,
    evaluationSet: "train" | "dev" | "test"
  ): Observable<ImageWithLabels[]> {
    return this.http
      .get<
        ImageWithLabels[]
      >(`${this.actionUrl}${modelId}/${evaluationSet}/all-images`)
      .pipe(catchError(this.handleError));
  }

  /**
   * Handle HTTP errors
   * @param error The HTTP error response
   * @returns Observable that throws an error
   */
  private handleError(error: HttpErrorResponse): Observable<never> {
    let errorMessage = "An error occurred";

    if (error.error instanceof ErrorEvent) {
      // Client-side or network error
      errorMessage = `Error: ${error.error.message}`;
    } else {
      // Backend error
      if (error.status === 404) {
        errorMessage = "Model or data not found";
      } else if (error.status === 400) {
        errorMessage = "Invalid request parameters";
      } else if (error.error && error.error.message) {
        errorMessage = error.error.message;
      } else {
        errorMessage = `Server error: ${error.status} - ${error.statusText}`;
      }
    }

    console.error("ConfusionMatrixService Error:", errorMessage);
    return throwError(() => new Error(errorMessage));
  }
}
